src/AzureRenderManager/WebApp/CostManagement/CostManagementClient.cs

```diff
@@ -0,0 +1,284 @@
+    }
+
+    public abstract class FilterExpression
+    {
+        public static FilterExpression Or(params FilterExpression[] expressions)
+            => new OrExpression(expressions);
+
+        public static FilterExpression And(params FilterExpression[] expressions)
+            => new AndExpression(expressions);
+
+        public static FilterExpression Dimension(string name, Operator op, IEnumerable<string> values)
+            => new DimensionExpression(new OperatorExpression(name, op, values));
+
+        public static FilterExpression Tag(string name, Operator op, IEnumerable<string> values)
+            => new TagExpression(new OperatorExpression(name, op, values));
+
+        private class AndExpression : FilterExpression
+        {
+            public AndExpression(params FilterExpression[] and)
+            {
+                And = and.ToArray();
+            }
+
+            public new IReadOnlyList<FilterExpression> And { get; }
+        }
+
+        private class OrExpression : FilterExpression
+        {
+            public OrExpression(params FilterExpression[] or)
+            {
+                Or = or.ToArray();
+            }
+
+            public new IReadOnlyList<FilterExpression> Or { get; }
+        }
+
+        private class TagExpression : FilterExpression
+        {
+            public TagExpression(OperatorExpression expression)
+                => Tags = expression;
+
+            public OperatorExpression Tags { get; }
+        }
+
+        private class DimensionExpression : FilterExpression
+        {
+            public DimensionExpression(OperatorExpression expression)
+                => Dimension = expression;
+
+            public new OperatorExpression Dimension { get; }
+        }
+
+        private class OperatorExpression
+        {
+            public OperatorExpression(string name, Operator op, IEnumerable<string> values)
+            {
+                Name = name;
+                Operator = op;
+                Values = values.ToList();
+            }
+
+            public string Name { get; }
+
+            [JsonConverter(typeof(StringEnumConverter))]
+            public Operator Operator { get; }
+
+            public IReadOnlyList<string> Values { get; }
+        }
+    }
+
+    public enum Timeframe
+    {
+        MonthToDate,
+        YearToDate,
+        WeekToDate,
+    }
+
+    public enum Granularity
+    {
+        Daily
+    }
+
+    public enum Operator
+    {
+        Contains,
+        In
+    }
+
+    public class CostManagementClient 
+    {
+        private readonly HttpClient _client;
+        private readonly string _accessToken;
+
+        public CostManagementClient(HttpClient client, string accessToken)
+        {
+            _client = client;
+            _accessToken = accessToken;
+        }
+
+        private static Uri GetUri(string scope)
+            => new Uri($"https://management.azure.com/{scope}/providers/Microsoft.CostManagement/query?api-version=2019-01-01");
```

Comment:
I thought scope started with a '/'?

Reply:
It does - we should add a check to reduce multiple slashes.

Reply:
Yes, will remove the leading slash. It still works though 🙂